zerver/lib/push_notifications.py

```diff
@@ -1032,16 +1032,31 @@ def get_message_payload(
             data["channel_id"] = channel_id
 
         data["topic"] = get_topic_display_name(message.topic_name(), user_profile.default_language)
-    elif message.recipient.type == Recipient.DIRECT_MESSAGE_GROUP:
-        data["recipient_type"] = "private" if for_legacy_clients else "direct"
-        # For group DMs, we need to fetch the users for the pm_users field.
-        # Note that this doesn't do a separate database query, because both
-        # functions use the get_display_recipient_by_id cache.
-        recipients = get_display_recipient(message.recipient)
-        if len(recipients) > 2:
-            data["pm_users"] = direct_message_group_users(message.recipient.id)
-    else:  # Recipient.PERSONAL
-        data["recipient_type"] = "private" if for_legacy_clients else "direct"
+    else:
+        assert message.recipient.type in [Recipient.PERSONAL, Recipient.DIRECT_MESSAGE_GROUP]
+        if for_legacy_clients:
+            data["recipient_type"] = "private"
+            if message.recipient.type == Recipient.DIRECT_MESSAGE_GROUP:
+                # For group DMs, we need to fetch the users for the pm_users field.
+                # Note that this doesn't do a separate database query, because both
+                # functions use the get_display_recipient_by_id cache.
+                recipients = get_display_recipient(message.recipient)
+                if len(recipients) > 2:
+                    data["pm_users"] = direct_message_group_users(message.recipient.id)
+        else:
+            data["recipient_type"] = "direct"
+            # For DMs and group DMs, we need to fetch the users for the `recipient_user_ids`
+            # field. Note that this doesn't do a separate database query, because it uses
+            # the `get_display_recipient_by_id` cache.
+            display_recipients = get_display_recipient(message.recipient)
+            recipient_user_ids = [
+                display_recipient["id"] for display_recipient in display_recipients
+            ]
+            if len(display_recipients) == 1:  # Recipient.PERSONAL
+                # `recipient_user_ids` should include both user IDs
+                # participating in 1:1 conversation.
+                recipient_user_ids.append(message.sender.id)
+            data["recipient_user_ids"] = sorted(recipient_user_ids)
```

Comment:
For a self-1:1, this will duplicate the user, right?

Let's add a test for that case :slightly_smiling_face:

Reply:
If self-1:1, there'd be no event received by the `missedmessage_mobile_notifications` worker in the first place.

Reply:
Is that certain?

If that is an assumption this code is making, it'd be good to make that explicit with an assert.

Reply:
I'll add an assert & a test.

Reply:
Added an assert.

```py
            if len(recipient_user_ids) == 1:  # Recipient.PERSONAL
                # `recipient_user_ids` should include both user IDs
                # in the 1:1 conversation.
                sender_id = message.sender.id
                # This won't cause duplication in the case of self DMs
                # as no push notification is involved in that case.
                assert recipient_user_ids[0] != sender_id
                recipient_user_ids.append(sender_id)
 ```

We already have a test `test_is_push_notifiable` which verifies that no push notif is sent in the self DM case.